### PR TITLE
Make `-q` to silence header, and `stdout` if specified twice

### LINKED ISSUE #4998: Suppress `godot` output, but not script output with `-q`
### Describe the project you are working on

Command line script that uses Godot for visualization and stdout for piping results.

### Describe the problem or limitation you are having in your project

`stdout` from script is always prepended with Godot version and debug info.
```
Godot Engine v3.4.4.stable.fedora.419e713a2 - https://godotengine.org
OpenGL ES 3.0 Renderer: Mesa Intel(R) HD Graphics 4400 (HSW GT2)
OpenGL ES Batching: ON
```
Using `--quiet` removes all `stdout` messages, including those from script,
and I'd to still read script messages.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow to specify `--quiet` (or `-q` in 4.0) multiple times.

`-q` - silences `godot` messages, but leaves script output
`-q -q` - silences all `stdout` output

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Pseudo-code.

https://github.com/godotengine/godot/pull/63575/files

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not from within Godot.

### Is there a reason why this should be core and not an add-on in the asset library?

Addons don't have control over internal Godot messages.

### PR DESCRIPTION
**Pull request by abitrolly**
I am not sure it works, because `quiet_header` is defined in `Main::setup` and probably is not available in `Main::setup2`.

I'd like to silence the header, because I use some `.gd` scripts from command line, and need to pipe their output without extra lines from `godot`.

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/4998.*

### CODE REVIEW

File: main/main.cpp [hunk 1]
```diff
@@ -712,8 +713,12 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 
 			OS::get_singleton()->_verbose_stdout = true;
 		} else if (I->get() == "-q" || I->get() == "--quiet") { // quieter output
-
-			quiet_stdout = true;
+			if quiet_header {
```
Comment (Spartan322) [inline main/main.cpp:716]: This isn't a valid syntax btw.

Comment (abitrolly) [reply main/main.cpp:716]: Yea. I still didn't setup the environment to compile `godot`, so I hoped for CI to do this part.

Comment (abitrolly) [reply main/main.cpp]: Fixed.

**Comment by Calinou** [comment]
`-qq` should probably also be supported as an equivalent of `-q -q`.

**Review by akien-mga — CHANGES_REQUESTED** [review]
I'm not fond of changing the meaning of `-q` and requiring specifying it twice to get back the proper `quiet` mode. I know some CLI works like this but that's bad design IMO (and usually it's for stuff like `-v`/`-vv` where it increases the verbosity, but here the `-q` behavior as of this PR does not make the output "quiet".

Instead I would add a `--no-header` parameter to silence the header independently from the rest.

**Comment by abitrolly** [comment]
@akien-mga `--no-header` will only turn off this line.
```
Godot Engine v3.4.5.stable.fedora.f9ac000d5 - https://godotengine.org
```
But there are also these that are output in default mode.
```
OpenGL ES 3.0 Renderer: Mesa Intel(R) HD Graphics 4400 (HSW GT2)
OpenGL ES Batching: ON
```
Should they also depend on `--no-header`?

Can you tell what else would `godot` print in default mode? My use case is to make `stdout` completely clear from `godot` messages, but not from script messages to `stdout`. If `--no-header` is enough for that, I can redo PR for the new option.

**Comment by Calinou** [comment]
> Should they also depend on --no-header?

I think so, yes.

**Comment by Calinou** [comment]
> I hope they are printed to stderr, no?

They should be if I'm not mistaken :slightly_smiling_face:

**Comment by abitrolly** [comment]
```
OpenGL ES 3.0 Renderer: Mesa Intel(R) HD Graphics 4400 (HSW GT2)
OpenGL ES Batching: ON
```
For these two lines I still need a pointer to filter them out.

**Comment by abitrolly** [comment]
Seems doable. I still haven't found the class that outputs this info though.

**Comment by Calinou** [comment]
In the Vulkan renderer, the header is printed here: https://github.com/Calinou/godot/blob/d5e2e8b0fc0e48c18ce80b92cef88c131a0d1c4b/drivers/vulkan/vulkan_context.cpp#L1033-L1035

In the OpenGL renderer, the header is printed here: https://github.com/Calinou/godot/blob/d5e2e8b0fc0e48c18ce80b92cef88c131a0d1c4b/drivers/gles3/rasterizer_gles3.cpp#L196

**Comment by kaissouDev** [comment]
But then you'll have to change the way "disable stdout" works or add a new parameter?

**Comment by Calinou** [comment]
> But then you'll have to change the way "disable stdout" works or add a new parameter?

Yes, a separate project setting should be added to toggle the header display.

**Comment by abitrolly** [comment]
I reworked this in https://github.com/godotengine/godot/pull/79179 to add `--no-header` option. Please take a look.

**Comment by kaissouDev** [comment]
Good work !

**Comment by kaissouDev** [comment]
@abitrolly I opened a pull request based on tient to add --no-header to the project settings

**Comment by abitrolly** [comment]
@kaissouDev nice! What is your story for exporting the game without console output? Are you making a roguelike for terminal in Godot? My use case for terminal is just have a script with a `#!/usr/bin/env godot --no-header -q` shebang and run it when needed for data processing and vis in a popup window.

**Comment by kaissouDev** [comment]
> @kaissouDev nice! What is your story for exporting the game without console output? Are you making a roguelike for terminal in Godot? My use case for terminal is just have a script with a `#!/usr/bin/env godot --no-header -q` shebang and run it when needed for data processing and vis in a popup window.

Hello ! I just wanted to add the "--quiet-header" option to have more control over the console of our game without having 2 lines written automatically.

**Comment by abitrolly** [comment]
Maybe an option to customize that header would be a better fit for project options? The it can

1. show the default engine info (advertise Godot engine)
2. be set empty
3. show customized engine info (and if the link to company support page)

But I think that discussion belongs to a new issue at https://github.com/godotengine/godot-proposals

**Comment by kaissouDev** [comment]
Yes, that would be a good idea, I think it can be done if you store a variable and print it in the console, but then I don't really see the point because you can use gdscript for everything.

**Comment by abitrolly** [comment]
Yea, well, I guess I am just overthinking it. That header reminded me how Unity logo flashes on Kongregate games. Guess that's a feature of free Unity version.

**Comment by Calinou** [comment]
I think that if you want to print a custom header, you should do it yourself in GDScript in an autoload's `_ready()` function. This allows using `print_rich()` and the like, something that a built-in implementation may not be able to do.

**Comment by kaissouDev** [comment]
> you should do it yourself in GDScript in an autoload's _ready() function. This allows using print_rich()

but isn't print_rich for making texts in different colours?

**Comment by Calinou** [comment]
> but isn't print_rich for making texts in different colours?

Yes, and some people may want to use that in a header :slightly_smiling_face:

**Comment by kaissouDev** [comment]
yes, I think it's preferable to use gdscript rather than integrate it directly into the editor, as that would take more time and be pointless

**Comment by akien-mga** [comment]
Superseded by #79179.

**Pull request closed without merging**